engine/storage/image/src/main/java/org/apache/cloudstack/storage/image/TemplateDataFactoryImpl.java

```diff
@@ -244,7 +246,7 @@ public TemplateInfo getReadyBypassedTemplateOnPrimaryStore(long templateId, Long
         if (poolId == null) {
             //Get ISO from existing pool ref
             HostVO host = hostDao.findById(hostId);
-            List<StoragePoolVO> pools = getStoragePoolsFromClusterOrZone(host.getClusterId(), host.getDataCenterId(), host.getHypervisorType());
+            List<StoragePoolVO> pools = getStoragePoolsFromClusterOrZoneOrLocal(host.getClusterId(), host.getDataCenterId(), host.getHypervisorType(), hostId);
```

Comment:
here we can 
```suggestion
            List<StoragePoolVO> pools = getStoragePoolsFromClusterOrZoneOrLocal(host.getClusterId(), host.getDataCenterId(), host.getHypervisorType(), hostId);
            if (pools.isEmpty()) {
                 throw new CloudRuntimeException("No storage pool found where to download template: " + templateId);
            }
```
because we know `null` will be returned otherwise. (and I think that should be an exception as well.)

Reply:
Refactored